libsolidity/codegen/LValue.cpp

```diff
@@ -313,9 +314,9 @@ void StorageItem::storeValue(Type const& _sourceType, SourceLocation const& _loc
 			solAssert(sourceType.location() != DataLocation::CallData, "Structs in calldata not supported.");
 			for (auto const& member: structType.members(nullptr))
 			{
-				// assign each member that is not a mapping
+				// assign each member that can live outside of storage
 				TypePointer const& memberType = member.type;
-				if (memberType->category() == Type::Category::Mapping)
+				if (!memberType->canLiveOutsideStorage())
```

Comment:
Hm, that also excludes copying of arrays of structs that contain mappings, right? I think they should still be copied, because there is still something to copy. On the other hand: What does the type checker say about these?

We plan to disallow such types being used in memory altogether with 0.6.0, so if this is too much work, it is probably not worth it.

Reply:
> Hm, that also excludes copying of arrays of structs that contain mappings, right?

I don't think so

Array's `canLiveOutsideStorage` depends on base type
https://github.com/ethereum/solidity/blob/9a00729ce750997920851a8a18c17d1797e852af/libsolidity/ast/Types.h#L708

Struct's `canLiveOutsideStorage` is always true.
https://github.com/ethereum/solidity/blob/9a00729ce750997920851a8a18c17d1797e852af/libsolidity/ast/Types.h#L836

So `arrayOfStructThatContainMappingsType.canLiveOutsideStorage()` should be true.

PS.
For each member of struct, it will recursively call `storeValue()`
https://github.com/ethereum/solidity/pull/6833/files/15e41ee4954aba6173ca5041bdf3f4f5b1e1ba4e#diff-bba7240285384c9f6e5edd7942acc957L346
That's why struct's `canLiveOutsideStorage()` can always return true.

Reply:
Ah that1s great! So I think nothing needs to be changed here.

Can you please add a test, though? I don't think semantic tests are overkill here - it is always good to also verify the semantics of code generation.

Please also add a test that uses an array of a struct that only contains a mapping - i'm curious because this more or less generates an empty struct.

Reply:
> On the other hand: What does the type checker say about these?

I didn't see any strange behavior.